Add secure software development guide (#986)
mattermost/mattermost-handbook: operations/research-and-development/product/development-process/secure-software-development.md

# Secure Software Development at Mattermost

This document provides guidelines to secure software development at Mattermost. The document is at the moment a [1% draft](/company/about-mattermost/mindsets#drafts-at-1-50-99) and not conclusive.

## Document scope

This document covers the secure software development practices on all source repositories used to create software products offered by Mattermost.

The product security team is overall in charge of secure software development practices by providing help and guidance and performing various verification activities, including code reviews and testing. Development teams are encouraged to adopt the secure software development practices described here. Third-party plugin developers are also welcome to use this document for guidance.

> ## Recommended reading

Reviewer: OWASP Top 10, Cheat Sheet Series, ..?

* [OSSF Scorecard Action Check Documentation](https://github.com/ossf/scorecard/blob/main/docs/checks.md)
* [Overview of Software Supply Chain Attacks](https://www.crowdstrike.com/cybersecurity-101/cyberattacks/supply-chain-attacks/) by [Crowdstrike](https://www.crowdstrike.com/)

## Code reviews

### Manual code reviews

The product security team performs reviews on new features and plugins when necessary. Here's some of the things included in the checks:

> * TODO.

Reviewer 2: ```suggestion
* Manual diff review
* Static application security testing (SAST) using multiple tools
* Software composition analysis (SCA)
* Dynamic application security testing (DAST)

Detailed review checklists are documented in the [security team shared drive](https://drive.google.com/drive/folders/1s3paBN6sfGRP-NHNuoekDHY_nrtRVagt), accessible to team members only.
```

### Automated code reviews

> The key Mattermost repositories have (or should have) automated code security analysis tools set up. In practice, we use [CodeQL](https://codeql.github.com/) as a Github Action to automatically analyze all pull requests. Additionally, the use of the [OSSF Security Scorecards GitHub Action](https://github.com/ossf/scorecard) is strongly encouraged. This document addresses many of the best practices checked by the Scorecards action.

Reviewer 2: Should we include a call to action in case a developer notices a repo that *doesn't* have security automation in place?

Reviewer: Let's remove `should have` and add a call to action. I'd also remove `key` and say `production` (vs. WIP Plugins)

## Dependency pinning

> Dependency pinning hardens the software development workflow against [software supply chain attacks](https://www.crowdstrike.com/cybersecurity-101/cyberattacks/supply-chain-attacks/). In this context, the specific threat being mitigated is upstream dependencies being taken over by attackers and replaced with malicious content. The risk commonly involves referring to dependencies without explicitly specifying a computed hash value for the dependency. In case of a dependency takeover, pulling the _latest_ version of it realizes the risk. The risk probability is considered low for supply chain attacks, but the effect can be pretty bad.

Reviewer 2: ```suggestion
Dependency pinning hardens the software development workflow against [software supply chain attacks](https://arstechnica.com/information-technology/2019/08/the-year-long-rash-of-supply-chain-attacks-against-open-source-is-getting-worse/). In this context, the specific threat being mitigated is upstream dependencies being taken over by attackers and replaced with malicious content. The risk commonly involves referring to dependencies without explicitly specifying a computed hash value for the dependency. In case of a dependency takeover, pulling the _latest_ version of it realizes the risk. The risk probability is considered low for supply chain attacks, but the effect can be pretty bad.
```

The link is a bit marketing-heavy. A news article like this one might do a better job at conveying the actual risks.

Practically, defining build-time dependencies with an explicit hash is the preferred solution. When specifying hashes to existing dependencies that did not use hashed references before, picking the hash of the currently used dependency is a good starting point as it is often not practical to do a full audit of all dependencies. Picking a hash and sticking to it doesn't ensure that said hash represents a safe version, it only ensures that a possibly malicious later version is not introduced in the build process afterwards.

> Please note also that pinning dependencies to a specific version (for example, `@2.1.1`) does not guard against malicious updates to the dependency. It is necessary to pin a dependency to a specific checksum (which is a cryptograhic guarantee to the uniqueness of the actual content).

Reviewer 2: This is debatable and potentially a little bit confusing. The version number alone isn't sufficient to pin a dependency in any ecosystem, that's true, but most dependency management tools will pin the hash automatically as long as you just specify a version. E.g. Go (automatic pinning in `go.sum`), npm (automatic pinning in `package-lock.json`), ... So in practice, a developer working with those ecosystems doesn't actually need to care about hashes.

Author: Good point, npm and Go are described later in the doc anyway

### Npm dependencies

> With npm, replace `npm install` with `npm ci` everywhere in the build/CI pipeline. This ensures that only packages matching `package-lock.json` are installed, see [this Stackoverflow answer](https://stackoverflow.com/questions/52499617/what-is-the-difference-between-npm-install-and-npm-ci) for more details and a reference to the npm documentation.

Reviewer 2: What's the value in linking to the SO answer instead of the docs page directly? Doesn't seem like the answer adds much and the docs are more likely to stay up to date.

> 

Reviewer 2: ```suggestion

When adding new dependencies, prefer [exact versions](https://docs.npmjs.com/cli/v8/commands/npm-install#save-exact) over the default semver range.

```

# Secure Software Development at Mattermost

This document provides guidelines to secure software development at Mattermost. The document is at the moment a [1% draft](/company/about-mattermost/mindsets#drafts-at-1-50-99) and not conclusive.

## Document scope

This document covers the secure software development practices on all source repositories used to create software products offered by Mattermost.

The product security team is overall in charge of secure software development practices by providing help and guidance and performing various verification activities, including code reviews and testing. Development teams are encouraged to adopt the secure software development practices described here. Third-party plugin developers are also welcome to use this document for guidance.

## Recommended reading

* [OSSF Scorecard Action Check Documentation](https://github.com/ossf/scorecard/blob/main/docs/checks.md)

> * [Overview of Software Supply Chain Attacks](https://www.crowdstrike.com/cybersecurity-101/cyberattacks/supply-chain-attacks/) by [Crowdstrike](https://www.crowdstrike.com/)

Reviewer 2: ```suggestion
```

Let's get rid of this one. It's the same link that got replaced with an Ars Technica article in the `Dependency pinning` section. It's otherwise a resonable link but the way it's written is a bit hyperbolic ("EVERY BUSINESS", emphasis theirs) and the section on prevention and detection is just pushing products.

* [OWASP Top Ten](https://owasp.org/www-project-top-ten/), a curated and updated list on most common security risks in web applications.
* [OWASP Application Security Verification Standard](https://owasp.org/www-project-application-security-verification-standard/)
* [OWASP Cheat Sheet Series](https://cheatsheetseries.owasp.org/)

## Code reviews

### Manual code reviews

The product security team performs reviews on new features and plugins when necessary. Here's some of the things included in the checks:

* Manual diff review
* Static application security testing (SAST) using multiple tools
* Software composition analysis (SCA)
* Dynamic application security testing (DAST)

Detailed review checklists are documented in the [security team shared drive](https://drive.google.com/drive/folders/1s3paBN6sfGRP-NHNuoekDHY_nrtRVagt), accessible to team members only.

### Automated code reviews

> The key Mattermost repositories have automated code security analysis tools set up. If this is not the case, please reach out to the Product Security Team for assistance in setting this up. Particular automated tools in each repository depend on the technologies used. In practice, we use [CodeQL](https://codeql.github.com/) as a Github Action to automatically analyze all pull requests. Additionally, the use of the [OSSF Security Scorecards GitHub Action](https://github.com/ossf/scorecard) is strongly encouraged. This document addresses many of the best practices checked by the Scorecards action.

Reviewer 2: ```suggestion
Mattermost production repositories have automated code security analysis tools set up. If this is not the case for a specific repository, please reach out to the product security team for assistance.

The automated tools in each repository depend on the technologies used in the repository. In practice, we use [CodeQL](https://codeql.github.com/) as a Github Action to automatically analyze all pull requests. Additionally, the use of the [OSSF Security Scorecards GitHub Action](https://github.com/ossf/scorecard) is strongly encouraged. This document describes some of the best practices checked by the Scorecards action.
```

Change `key` to `production` as suggested by @DSchalla earlier. Plus minor clarifying changes in wording.

## Dependency pinning

> Dependency pinning hardens the software development workflow against [software supply chain attacks](https://arstechnica.com/information-technology/2019/08/the-year-long-rash-of-supply-chain-attacks-against-open-source-is-getting-worse/). In this context, the specific threat being mitigated is upstream dependencies being taken over by attackers and replaced with malicious content. The risk commonly involves referring to dependencies without explicitly specifying a computed hash value for the dependency. In case of a dependency takeover, pulling the _latest_ version of it realizes the risk. The risk probability is considered low for supply chain attacks, but the effect can be pretty bad.

Reviewer 2: ```suggestion
Dependency pinning hardens software development workflows against [software supply chain attacks](https://arstechnica.com/information-technology/2019/08/the-year-long-rash-of-supply-chain-attacks-against-open-source-is-getting-worse/). The specific threat being mitigated is upstream dependencies being taken over by attackers and replaced with malicious content. Attacks are more likely to be successful against projects that refer to their dependencies without explicitly specifying a computed hash value for the dependency: In case of a dependency takeover, pulling the _latest_ version triggers the attack. The likelihood of this risk materializing is low, but the impact can be bad.
```

Minor wording suggestions.

> Practically, defining build-time dependencies with an explicit hash is the preferred solution. When specifying hashes to existing dependencies that did not use hashed references before, picking the hash of the currently used dependency is a good starting point as it is often not practical to do a full audit of all dependencies. Picking a hash and sticking to it doesn't ensure that said hash represents a safe version, it only ensures that a possibly malicious later version is not introduced in the build process afterwards. Some technologies support automatic dependency pinning so that the developer doesn't need to explicitly find and set dependency hashes, for example Go and npm, as described below.

Reviewer 2: ```suggestion
In practice, defining build-time dependencies with an explicit hash is the preferred solution. When specifying hashes to existing dependencies that did not use hash-based references before, picking the hash of the currently used dependency is a good starting point as it is often not practical to do a full audit of all dependencies. Picking a hash and sticking to it doesn't ensure that the hash represents a safe version, but it does ensures that a possibly malicious later version is not introduced in the build process afterwards.

Many dependency managers support automatic dependency pinning so that a developer doesn't need to explicitly find and set dependency hashes. For example both Go and npm do, as described below.
```

Minor wording suggestions.

### Npm dependencies

> With npm, replace `npm install` with `npm ci` everywhere in the build/CI pipeline. This ensures that only packages matching `package-lock.json` are installed, see [the documentation on `npm ci`](https://docs.npmjs.com/cli/v8/commands/npm-ci) for more details and a reference to the npm documentation.

Reviewer 2: ```suggestion
With npm, always use `npm ci` instead of `npm install` in build scripts and automated pipelines. This ensures that only packages matching the existing `package-lock.json` file are installed. See [the documentation on `npm ci`](https://docs.npmjs.com/cli/v8/commands/npm-ci) for more details.
```

Minor wording suggestions + remove the superfluous ` and a reference to the npm documentation`.

When adding new dependencies, prefer [exact versions](https://docs.npmjs.com/cli/v8/commands/npm-install#save-exact) over the default semver range.

### Docker

> Docker image dependencies exist in Dockerfiles and elsewhere, such as in CircleCI config files. The [Dockpin](https://github.com/Jille/dockpin) tool can be used to pin Docker base images to their currently latest versions. To pin a dockerfile, `dockpin docker pin -f Dockerfile`. To pin Docker base image dependencies elsewhere, `dockpin docker resolve [base-image]`, for example, `dockpin docker resolve ubuntu@20.04`. Here's a handy script to do this on CircleCI yaml files, replacing the image reference with one with a hash, and additionally commenting with the timestamp the hash was obtained:

Reviewer 2: ```suggestion
Docker image dependencies exist in Dockerfiles and elsewhere, such as in CircleCI config files. The [Dockpin](https://github.com/Jille/dockpin) tool can be used to pin Docker base images to their currently latest versions. To pin a dockerfile, run `dockpin docker pin -f Dockerfile`. To pin Docker base image dependencies elsewhere, run `dockpin docker resolve [base-image]`; for example `dockpin docker resolve ubuntu@20.04`.

Here's a handy script to do this on CircleCI yaml files, replacing the image reference with one with a hash, and additionally commenting with the timestamp the hash was obtained:
```

Minor wording suggestions.

```
#!/bin/bash

ymlfile=$1

if [ -z "$1" ] ; then
    echo usage: $0 circleci-config-ymlfile
    echo example: $0 ./.circleci/config.yml
    exit 0
fi

for x in `grep -A1 docker $1|grep -- '- image:'|sed 's/.*image:\ \(.*\)/\1/g'|sed "s/'//g"|sort|uniq`;do
    hashed=`dockpin docker resolve $x`
    echo base $x hashed $hashed
    sed -i "s~$x~$hashed\ #\ $x,\ `date`~g" $ymlfile
done
```

### Go

> When adding a Go dependency using `go get`, the dependency can be installed specifying either a specific version (`go get example.com/theirmodule@v1.3.4`), or the latest available version (`go get example.com/theirmodule@latest`), or a commit hash (`go get example.com/theirmodule@4cf76c2`) or a branchname (`@branchname`). In all cases, the used dependencies are listed in the automatically generated `go.mod` file and the content hashes of the dependencies are written to the `go.sum` file. Make sure to commit the `go.mod` and `go.sum` files to your repository. This ensures that subsequent builds download exactly the same dependencies.

Reviewer 2: ```suggestion
When adding a Go dependency using `go get`, the dependency can be installed specifying either a specific version (`go get example.com/theirmodule@v1.3.4`), the latest available version (`go get example.com/theirmodule@latest`), or a commit hash (`go get example.com/theirmodule@4cf76c2`) or branchname (`@branchname`). In all cases, the installed dependencies are added to the automatically generated `go.mod` file and the content hashes of the dependencies are written to the `go.sum` file. Make sure to commit both the `go.mod` and the `go.sum` file to your repository. This ensures that subsequent builds download exactly the same dependencies.
```

Minor wording suggestions.

### Github Actions

> GitHub Actions are specified in the `.yml` files in the `.github/workflows` directory inside the Github repository. Check [this nice tool](https://app.stepsecurity.io/) by [Stepsecurity](https://www.stepsecurity.io/) that can harden the workflow spec. Check at least the "Restrict permissions..." and "Pin actions..." boxes, then paste the workflow `.yml` file and click Secure workflow. Then, review the changes and paste the result back into the workflow `.yml` file.

Reviewer 2: ```suggestion
GitHub Actions are specified in `.yml` files in the `.github/workflows` directory inside a GitHub repository. Check out [this nice tool](https://app.stepsecurity.io/) by [Stepsecurity](https://www.stepsecurity.io/) that can harden the workflow spec. Check at least the "Restrict permissions..." and "Pin actions..." boxes, then paste in the workflow `.yml` file and click "Secure workflow". Review the changes and paste the result back into your workflow `.yml` file.
```

Minor wording suggestions.

# Secure Software Development at Mattermost

> This document provides guidelines to secure software development at Mattermost. The document is at the moment a [1% draft](/company/about-mattermost/mindsets#drafts-at-1-50-99) and not conclusive.

Reviewer: Would you say after this it's at 50% or 1% still correct? :-)

Author: lol good point i'd say we are at 50% now

When adding new dependencies, prefer [exact versions](https://docs.npmjs.com/cli/v8/commands/npm-install#save-exact) over the default semver range.

### Docker

Docker image dependencies exist in Dockerfiles and elsewhere, such as in CircleCI config files. The [Dockpin](https://github.com/Jille/dockpin) tool can be used to pin Docker base images to their currently latest versions. To pin a dockerfile, run `dockpin docker pin -f Dockerfile`. To pin Docker base image dependencies elsewhere, run `dockpin docker resolve [base-image]`; for example `dockpin docker resolve ubuntu@20.04`.

Here's a handy script to do this on CircleCI yaml files, replacing the image reference with one with a hash, and additionally commenting with the timestamp the hash was obtained:

```
#!/bin/bash

ymlfile=$1

if [ -z "$1" ] ; then
    echo usage: $0 circleci-config-ymlfile
    echo example: $0 ./.circleci/config.yml
    exit 0
fi

for x in `grep -A1 docker $1|grep -- '- image:'|sed 's/.*image:\ \(.*\)/\1/g'|sed "s/'//g"|sort|uniq`;do
    hashed=`dockpin docker resolve $x`
    echo base $x hashed $hashed
    sed -i "s~$x~$hashed\ #\ $x,\ `date`~g" $ymlfile
done
```

### Go

> When adding a Go dependency using `go get`, the dependency can be installed specifying either a specific version (`go get example.com/theirmodule@v1.3.4`), the latest available version (`go get example.com/theirmodule@latest`), or a commit hash (`go get example.com/theirmodule@4cf76c2`) or branchname (`@branchname`). In all cases, the installed dependencies are added to the automatically generated `go.mod` file and the content hashes of the dependencies are written to the `go.sum` file. Make sure to commit both the `go.mod` and the `go.sum` file to your repository. This ensures that subsequent builds download exactly the same dependencies.

Reviewer: Do we have any specific recommendation what approach would be recommended?

Author: i don't think this document can suggest a single best practice that applies in all cases

Reviewer: That makes sense, we promote pinning in most examples for good reasons, shouldn't we discourage then using `latest` or is there a reason to use it with Go?

Reviewer 2: Using `@latest` doesn't actually store a reference to `@latest` in `go.mod` or `go.sum` -- instead it resolves to whatever is the current latest version. Using it should be fine since the exact version still gets pinned as it would with any other version query.

Reviewer: Perfect, thank you @jupenur!
